demos/common/images_capture.h

```diff
@@ -0,0 +1,119 @@
+#include <stdexcept>
+#include <string>
+#include <memory>
+
+#ifdef _WIN32
+#include <os/windows/w_dirent.h>
+#else
+#include <dirent.h>
+#endif
+
+#include <opencv2/core/mat.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/videoio.hpp>
+
+class InvalidInput : public std::invalid_argument {
+public:
+    InvalidInput() : std::invalid_argument{""} {}
+};
+
+class ImagesCapture {
+public:
+    bool loop;
+
+    ImagesCapture(bool loop) : loop{loop} {}
+    virtual cv::Mat read() = 0;
+    virtual ~ImagesCapture() = default;
+};
+
+class ImreadWrapper : public ImagesCapture {
+public:
+    cv::Mat img;
+    bool canRead;
+
+    ImreadWrapper(const std::string &input, bool loop) : ImagesCapture{loop}, canRead{true} {
+        img = cv::imread(input);
+        if(!img.data) throw InvalidInput{};
+    }
+
+    cv::Mat read() override {
+        if (loop) return img.clone();
+        if (canRead) {
+            canRead = false;
+            return img.clone();
+        }
+        return cv::Mat{};
+    }
+};
+
+class DirReader : public ImagesCapture {
+    std::string input;
+    DIR *dir;
```

Comment:
While reading all the images in advance is a bad idea, you should read all the _names_ in advance, because images should be output in alphabetical order, and to do that, you need all the file names in advance.

(I think alphabetical order is needed, because a directory of images is likely to be a directory of video frames, and the best hope to arrange those correctly is to sort by name.)

Reply:
Updated